Gah, bad timing pushing to try:

```
[task 2017-06-14T11:21:01.421797Z] 11:21:01     INFO -  error[E0061]: this function takes 2 parameters but 3 parameters were supplied
[task 2017-06-14T11:21:01.422049Z] 11:21:01     INFO -     --> /home/worker/workspace/build/src/servo/components/style/gecko/values.rs:407:26
[task 2017-06-14T11:21:01.422302Z] 11:21:01     INFO -      |
[task 2017-06-14T11:21:01.422577Z] 11:21:01     INFO -  407 |                   set_name(gecko_value, atom!("none").into_addrefed(), pres_context);
[task 2017-06-14T11:21:01.422881Z] 11:21:01     INFO -      |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 2 parameters
[task 2017-06-14T11:21:01.423117Z] 11:21:01     INFO -      |
[task 2017-06-14T11:21:01.423371Z] 11:21:01     INFO -     ::: /home/worker/workspace/build/src/obj-firefox/toolkit/library/x86_64-unknown-linux-gnu/release/build/style-a2ebb78ccb6eb042/out/gecko/bindings.rs
[task 2017-06-14T11:21:01.423505Z] 11:21:01     INFO -      |
[task 2017-06-14T11:21:01.423677Z] 11:21:01     INFO -  908 | /     pub fn Gecko_SetCounterStyleToName(ptr: *mut CounterStylePtr,
[task 2017-06-14T11:21:01.423860Z] 11:21:01     INFO -  909 | |                                        name: *mut nsIAtom);
[task 2017-06-14T11:21:01.424040Z] 11:21:01     INFO -      | |___________________________________________________________- defined here
[task 2017-06-14T11:21:01.424263Z] 11:21:01     INFO -  error[E0061]: this function takes 2 parameters but 3 parameters were supplied
[task 2017-06-14T11:21:01.424430Z] 11:21:01     INFO -     --> /home/worker/workspace/build/src/servo/components/style/gecko/values.rs:410:26
```

https://treeherder.mozilla.org/#/jobs?repo=try&revision=2e59616696ec527441273f8e9afa92990d56fe59

--- a/components/style/properties/declaration_block.rs
+++ b/components/style/properties/declaration_block.rs
@@ -14,6 +14,7 @@ use parser::{ParserContext, log_css_error};
 use properties::animated_properties::AnimationValue;
 use selectors::parser::SelectorParseError;
 use shared_lock::Locked;
+use smallvec::SmallVec;
 use std::fmt;
 use std::slice::Iter;
 use style_traits::{PARSING_MODE_DEFAULT, ToCss, ParseError, ParsingMode, StyleParseError};
@@ -605,24 +606,31 @@ impl ToCss for PropertyDeclarationBlock {
             // Step 3.3
             let shorthands = declaration.shorthands();
             if !shorthands.is_empty() {
-                // Step 3.3.1
-                //
-                // FIXME(emilio): All this looks terribly inefficient...
-                let mut longhands = self.declarations.iter()
-                    .filter(|d| !already_serialized.contains(d.0.id()))
-                    .collect::<Vec<_>>();
+                // Step 3.3.1 is done by checking already_serialized while
+                // iterating below.
 
                 // Step 3.3.2
                 for &shorthand in shorthands {
                     let properties = shorthand.longhands();
 
                     // Substep 2 & 3
-                    let mut current_longhands = Vec::new();
+                    let mut current_longhands = SmallVec::<[_; 10]>::new();
                     let mut important_count = 0;
                     let mut found_system = None;
 
-                    if shorthand == ShorthandId::Font && longhands.iter().any(|&&(ref l, _)| l.get_system().is_some()) {
-                        for &&(ref longhand, longhand_importance) in longhands.iter() {
+                    let is_system_font =
+                        shorthand == ShorthandId::Font &&
+                        self.declarations.iter().any(|&(ref l, _)| {
+                            !already_serialized.contains(l.id()) &&
+                            l.get_system().is_some()
+                        });
+
+                    if is_system_font {
+                        for &(ref longhand, longhand_importance) in &self.declarations {
+                            if already_serialized.contains(longhand.id()) {
+                                continue;
+                            }
+
                             if longhand.get_system().is_some() || longhand.is_default_line_height() {
                                 current_longhands.push(longhand);
                                 if found_system.is_none() {
@@ -634,7 +642,11 @@ impl ToCss for PropertyDeclarationBlock {
                             }
                         }
                     } else {
-                        for &&(ref longhand, longhand_importance) in longhands.iter() {
+                        for &(ref longhand, longhand_importance) in &self.declarations {
+                            if already_serialized.contains(longhand.id()) {
+                                continue;
+                            }
+
                             if longhand.id().is_longhand_of(shorthand) {
                                 current_longhands.push(longhand);
                                 if longhand_importance.important() {
@@ -728,11 +740,6 @@ impl ToCss for PropertyDeclarationBlock {
                     for current_longhand in &current_longhands {
                         // Substep 9
                         already_serialized.insert(current_longhand.id());
-                        let index_to_remove = longhands.iter().position(|l| l.0 == **current_longhand);
-                        if let Some(index) = index_to_remove {
-                            // Substep 10
-                            longhands.remove(index);
-                        }
                     }
                 }
             }

--- a/components/style/properties/properties.mako.rs
+++ b/components/style/properties/properties.mako.rs
@@ -551,6 +551,42 @@ impl LonghandId {
         }
     }
 
+    fn shorthands(&self) -> &'static [ShorthandId] {
+        // first generate longhand to shorthands lookup map
+        //
+        // NOTE(emilio): This currently doesn't exclude the "all" shorthand. It
+        // could potentially do so, which would speed up serialization
+        // algorithms and what not, I guess.
+        <%
+            longhand_to_shorthand_map = {}
+            for shorthand in data.shorthands:
+                for sub_property in shorthand.sub_properties:
+                    if sub_property.ident not in longhand_to_shorthand_map:
+                        longhand_to_shorthand_map[sub_property.ident] = []
+
+                    longhand_to_shorthand_map[sub_property.ident].append(shorthand.camel_case)
+
+            for shorthand_list in longhand_to_shorthand_map.itervalues():
+                shorthand_list.sort()
+        %>
+
+        // based on lookup results for each longhand, create result arrays
+        % for property in data.longhands:
+            static ${property.ident.upper()}: &'static [ShorthandId] = &[
+                % for shorthand in longhand_to_shorthand_map.get(property.ident, []):
+                    ShorthandId::${shorthand},
+                % endfor
+            ];
+        % endfor
+
+        match *self {
+            % for property in data.longhands:
+                LonghandId::${property.camel_case} => ${property.ident.upper()},
+            % endfor
+        }
+    }
+
+
     /// If this is a logical property, return the corresponding physical one in the given writing mode.
     /// Otherwise, return unchanged.
     pub fn to_physical(&self, wm: WritingMode) -> Self {
@@ -885,7 +921,7 @@ impl<'a> PropertyDeclarationId<'a> {
             PropertyDeclarationId::Longhand(id) => {
                 match *other {
                     PropertyId::Longhand(other_id) => id == other_id,
-                    PropertyId::Shorthand(shorthand) => shorthand.longhands().contains(&id),
+                    PropertyId::Shorthand(shorthand) => self.is_longhand_of(shorthand),
                     PropertyId::Custom(_) => false,
                 }
             }
@@ -899,7 +935,7 @@ impl<'a> PropertyDeclarationId<'a> {
     /// shorthand.
     pub fn is_longhand_of(&self, shorthand: ShorthandId) -> bool {
         match *self {
-            PropertyDeclarationId::Longhand(ref id) => shorthand.longhands().contains(id),
+            PropertyDeclarationId::Longhand(ref id) => id.shorthands().contains(&shorthand),
             _ => false,
         }
     }
@@ -1308,40 +1344,9 @@ impl PropertyDeclaration {
 
     /// The shorthands that this longhand is part of.
     pub fn shorthands(&self) -> &'static [ShorthandId] {
-        // first generate longhand to shorthands lookup map
-        <%
-            longhand_to_shorthand_map = {}
-            for shorthand in data.shorthands:
-                for sub_property in shorthand.sub_properties:
-                    if sub_property.ident not in longhand_to_shorthand_map:
-                        longhand_to_shorthand_map[sub_property.ident] = []
-
-                    longhand_to_shorthand_map[sub_property.ident].append(shorthand.camel_case)
-
-            for shorthand_list in longhand_to_shorthand_map.itervalues():
-                shorthand_list.sort()
-        %>
-
-        // based on lookup results for each longhand, create result arrays
-        % for property in data.longhands:
-            static ${property.ident.upper()}: &'static [ShorthandId] = &[
-                % for shorthand in longhand_to_shorthand_map.get(property.ident, []):
-                    ShorthandId::${shorthand},
-                % endfor
-            ];
-        % endfor
-
-        match *self {
-            % for property in data.longhands:
-                PropertyDeclaration::${property.camel_case}(_) => ${property.ident.upper()},
-            % endfor
-            PropertyDeclaration::CSSWideKeyword(id, _) |
-            PropertyDeclaration::WithVariables(id, _) => match id {
-                % for property in data.longhands:
-                    LonghandId::${property.camel_case} => ${property.ident.upper()},
-                % endfor
-            },
-            PropertyDeclaration::Custom(_, _) => &[]
+        match self.id() {
+            PropertyDeclarationId::Longhand(id) => id.shorthands(),
+            PropertyDeclarationId::Custom(..) => &[],
         }
     }
 

